src/L4LSP.hs

```diff
@@ -242,6 +242,12 @@ posInRange (Position line col) (RealSRng (SRng (Pos top left) (Pos bottom right)
   (line == top && col >= left || line > top)
   && (line == bottom && col <= right || line < bottom)
 
+-- | Converts 0-based SRng to 1-based SRng (ratchet but it works. we welcome any changes)
+incrementSRng :: SRng -> SRng
+incrementSRng (RealSRng (SRng (Pos startline startcol) (Pos endline endcol))) =
+  RealSRng (SRng (Pos (startline + 1) (startcol + 1)) (Pos (endline + 1) (endcol + 1)))
+incrementSRng x = x
+
```

Comment:
Having a function like this is kind of error prone, since we don't know if it is used at all the correct locations.

I think we should use different  types for the 0-indexed and 1-indexed location. We should probably not use the `lsp` version of ranges at all, other than when communicating with the lsp client. After that we would immediately convert to our own version of `Range`.

I think this was the case in some earlier version of the code.

Reply:
Or actually, this is already the case. What we need to do is to make the conversion from 1-indexed to 0-indexed in this function:
```
sRngToRange :: SRng -> Maybe Range
sRngToRange (RealSRng (SRng (Pos l1 c1) (Pos l2 c2))) = Just $ Range (Position l1 c1) (Position l2 c2)
sRngToRange (DummySRng _) = Nothing
```

Reply:
And also make the fix to this function:

```
posInRange :: Position -> SRng -> Bool
posInRange (Position _line _col) (DummySRng _) = False
posInRange (Position line col) (RealSRng (SRng (Pos top left) (Pos bottom right))) =
  (line == top && col >= left || line > top)
  && (line == bottom && col <= right || line < bottom
```

Reply:
The easiest way to fix the `posInRange` function without code duplication would be to call
`sRngToRange` on the `SRng` it gets.

The other alternative would be to call `positionToPos :: Position -> Pos` (which would convert 0-indexed to 1-indexed) before calling `posInRange`. The key here is that both arguments of `posInRange` should have the same kind of index (1-indexed (`SRng` and `Pos`) or 0-indexed (`Range` and `Position`).